File: Source/ShadcnBlazor/Components/ScrollArea/ScrollAreaScrollbarImpl.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;
using Microsoft.JSInterop;
using System;
using System.Linq;

namespace ShadcnBlazor;

public partial class ScrollAreaScrollbarImpl : ShadcnJSComponentBase
{
    DotNetObjectReference<ScrollAreaScrollbarImpl>? _dotnet;
    DomRect? _rect;

    public ScrollAreaScrollbarImpl() : base(
        JSModulePathBuilder.GetComponentPath("ScrollArea", "ScrollAreaScrollbarImpl"))
    { Id = Identifier.NewId(); }

    ElementReference? Scrollbar => RootContext?.ViewPort;

    [JSInvokable]
    public async Task HandleSizeChangeAsync()
    {
        if (Scrollbar == null)
        {
            return;
        }

        var viewportInfo = await GetElementInfoAsync(RootContext?.ViewPort);
        var scrollbarInfo = await GetElementInfoAsync(Scrollbar);

        var paddingStart = await GetPropertyValueAsync(Scrollbar, "padding-left");
        var paddingEnd = await GetPropertyValueAsync(Scrollbar, "padding-right");

        if (viewportInfo == null || scrollbarInfo == null)
        {
            return;
        }

        var size = new ScrollSize
        {
            Content = IsHorizontal ? viewportInfo.ScrollWidth : viewportInfo.ScrollHeight,
            ViewPort = IsHorizontal ? viewportInfo.OffsetWidth : viewportInfo.OffsetHeight,
            Scrollbar =
                new Scrollbar
                {
                    Size = IsHorizontal ? scrollbarInfo.ClientWidth : scrollbarInfo.ClientHeight,
                    PaddingStart = paddingStart,
                    PaddingEnd = paddingEnd
                }
        };
        Console.WriteLine($"HandleSizeChangeAsync: {size}");

        ScrollbarVisibleContext?.HandleSizeChange(size);
    }

    [JSInvokable]
    public void HandleWheel(WheelEventArgs args)
    {
        var maxScrollPos = ScrollbarVisibleContext?.Sizes?.Content - ScrollbarVisibleContext?.Sizes?.ViewPort;
        ScrollbarVisibleContext?.HandleWheelScroll(args, maxScrollPos ?? 0);
    }

    IJSObjectReference? _resizeObserver;

    ElementReference? ViewPort => RootContext?.ViewPort;

    ElementReference? Content => RootContext?.Content;

    protected override async ValueTask OnAfterImportAsync()
    {
        _dotnet ??= DotNetObjectReference.Create(this);
        await InvokeVoidAsync("addWheelListener", Scrollbar, _dotnet, nameof(HandleWheel));
        _resizeObserver = await InvokeAsync<IJSObjectReference>(
            "addResizeObserver",
            ViewPort,
            Scrollbar,
            Content,
            IsHorizontal,
            _dotnet,
            nameof(HandleSizeChangeAsync));
    }

    protected override async ValueTask OnDisposingAsync()
    {
        await InvokeVoidAsync("removeWheelListener");
        if (_resizeObserver != null)
        {
            await _resizeObserver.InvokeVoidAsync("dispose");
            await  _resizeObserver.DisposeAsync();
        }
    }

    async Task<DomRect> GetBoundingClientRectAsync(ElementReference element)
    { return await InvokeAsync<DomRect>("getBoundingClientRect", element); }
    async Task<ElementInfo> GetElementInfoAsync(ElementReference? element)
    { return await InvokeAsync<ElementInfo>("getElementInfo", element); }
    async Task<int> GetPropertyValueAsync(ElementReference? element, string prop)
    { return await InvokeAsync<int>("getComputedStyle", element, prop); }

    void HandleDragScroll(MouseEventArgs args)
    {
        if (_rect != null)
        {
            var x = args.ClientX - _rect.Left;
            var y = args.ClientY - _rect.Top;
            if (OnDragScroll.HasDelegate)
            {
                OnDragScroll.InvokeAsync((x, y));
            }
        }
    }

    async Task HandlePointerDown(ExtendedPointerEventArgs args)
    {
        var mainPointer = 0;
        if (args.Button == mainPointer)
        {
            var element = args.Target;
            await SetPointerCaptureAsync(element, args.PointerId);
            var rect = await GetBoundingClientRectAsync(Ref);
            if (rect != null)
            {
                _rect = rect;
            }
            if (Scrollbar != null)
            {
                await SetElementStyleAsync(Scrollbar, "scroll-behavior", "auto");
            }
            HandleDragScroll(args);
        }
    }

    void HandlePointerMove(ExtendedPointerEventArgs args) { HandleDragScroll(args); }

    async Task HandlePointerUp(ExtendedPointerEventArgs args)
    {
        var element = args.Target;
        await ReleasePointerCaptureasyn(element, args.PointerId);
        await SetElementStyleAsync(Scrollbar, "scroll-behavior", string.Empty);
        _rect = null;
    }

    async Task ReleasePointerCaptureasyn(string? id, long pointerId)
    { await InvokeVoidAsync("releasePointerCapture", id, pointerId); }
    async Task SetElementStyleAsync(ElementReference? element, string property, string value)
    { await InvokeVoidAsync("setElementStyle", element, property, value); }
    async Task SetPointerCaptureAsync(string? id, long pointerId)
    { await InvokeVoidAsync("setPointerCapture", id, pointerId); }

    record ElementInfo
    {
        public double ClientHeight { get; set; }

        public double ClientWidth { get; set; }

        public double OffsetHeight { get; set; }

        public double OffsetWidth { get; set; }

        public double ScrollHeight { get; set; }

        public double ScrollWidth { get; set; }
    }
}

File: Source/ShadcnBlazor/Components/ScrollArea/ScrollAreaScrollbarImpl.razor.js
export function setPointerCapture(id, pointerId) {
  const element = document.getElementById(id)
  if (element) {
    element.setPointerCapture(pointerId)
  }
}

export function getBoundingClientRect(element) {
  if (element) {
    return element.getBoundingClientRect()
  }
}

export function setElementStyle(element, property, value) {
  if (element) {
    element.style.setProperty(property, value)
  }
}

export function releasePointerCapture(id, pointerId) {
  const element = document.getElementById(id)
  if (element) {
    if (element.hasPointerCapture(pointerId)) {
      element.releasePointerCapture(pointerId)
    }
  }
}
let handleWheel = null
export function addWheelListener(scrollbar, dotNetHelper, method) {
  handleWheel = (event) => {
    const element = event.target
    const isScrollbarWheel = scrollbar.contains(element)
    if (isScrollbarWheel) {
      dotNetHelper.invokeMethodAsync(method, event)
    }
  }
  document.addEventListener('wheel', handleWheel)
}

export function removeWheelListener() {
  if (handleWheel) {
    document.removeEventListener('wheel', handleWheel)
    handleWheel = null
  }
}
function toInt(value) {
  return value ? Number.parseInt(value, 10) : 0
}

let prevState = null
export function addResizeObserver(viewportEl, scrollbarEl, contentEl, isHorizontal, dotnet, method) {
  if (!viewportEl || !scrollbarEl || !contentEl || !dotnet || !method)
    return null


  const areEqual = (a, b) => {
    if (!a || !b) return false
    return (
      a.content === b.content &&
      a.viewport === b.viewport &&
      a.scrollbar.size === b.scrollbar.size &&
      a.scrollbar.paddingStart === b.scrollbar.paddingStart &&
      a.scrollbar.paddingEnd === b.scrollbar.paddingEnd
    )
  }

  const handler = () => {
    const content = isHorizontal
      ? viewportEl.scrollWidth ?? 0
      : viewportEl.scrollHeight ?? 0

    const viewport = isHorizontal
      ? viewportEl.offsetWidth ?? 0
      : viewportEl.offsetHeight ?? 0

    const scrollbar = isHorizontal
      ? {
        size: scrollbarEl.clientWidth ?? 0,
        paddingStart: toInt(getComputedStyle(scrollbarEl).paddingLeft),
        paddingEnd: toInt(getComputedStyle(scrollbarEl).paddingRight)
      }
      : {
        size: scrollbarEl.clientHeight ?? 0,
        paddingStart: toInt(getComputedStyle(scrollbarEl).paddingTop),
        paddingEnd: toInt(getComputedStyle(scrollbarEl).paddingBottom)
      }

    const currentState = { content, viewport, scrollbar }

    if (areEqual(prevState, currentState))
      return

    prevState = currentState

    dotnet.invokeMethodAsync(method, currentState)
      .catch(() => { })
  }

  const scrollbarResizeObserver = new ResizeObserver(handler)
  const contentResizeObserver = new ResizeObserver(handler)

  scrollbarResizeObserver.observe(scrollbarEl)
  contentResizeObserver.observe(contentEl)

  return {
    dispose: () => {
      scrollbarResizeObserver.disconnect()
      contentResizeObserver.disconnect()
      prevState = null
    }
  }
}

export function getElementInfo(element) {
  if (element) {
    return {
      scrollWidth: element.scrollWidth,
      scrollHeight: element.scrollHeight,
      offsetWidth: element.offsetWidth,
      offsetHeight: element.offsetHeight,
      clientWidth: element.clientWidth,
      clientHeight: element.clientHeight,
    }
  }
  return null
}
export function getComputedStyle(element, prop) {
  if (element) {
    const value = toInt(window.getComputedStyle(element).getPropertyValue(prop))
    return value
  }
  return null
}